python/cugraph_service/cugraph_service_client/remote_graph.py

```diff
@@ -0,0 +1,530 @@
+    cannot be found, which allows for code to import optional dependencies, and
+    have only the code paths that use the module affected.
+    """
+
+    def __init__(self, mod_name):
+        self.name = mod_name
+
+    def __getattr__(self, attr):
+        raise RuntimeError(f"This feature requires the {self.name} " "package/module")
+
+
+try:
+    cudf = importlib.import_module("cudf")
+except ModuleNotFoundError:
+    cudf = MissingModule("cudf")
+
+
+class RemoteGraph:
+    def __init__(self, cgs_client, cgs_graph_id):
+        self.__client = cgs_client
+        self.__cgs_graph_id = cgs_graph_id
+
+    def is_remote(self):
+        return True
+
+    def is_bipartite(self):
+        return False
+
+    def is_multipartite(self):
+        return False
+
+    def is_directed(self):
+        return True
+
+    def is_multigraph(self):
+        return True
+
+    def is_weighted(self):
+        return True
+
+    def has_isolated_vertices(self):
+        raise NotImplementedError("not implemented")
+
+    def to_directed(self):
+        raise NotImplementedError("not implemented")
+
+    def to_undirected(self):
+        raise NotImplementedError("not implemented")
+
+    @property
+    def edgelist(self):
+        raise NotImplementedError("not implemented")
+
+    @property
+    def adjlist(self):
+        raise NotImplementedError("not implemented")
+
+
+class RemotePropertyGraph:
+    # column name constants used in internal DataFrames
+    vertex_col_name = "_VERTEX_"
+    src_col_name = "_SRC_"
+    dst_col_name = "_DST_"
+    type_col_name = "_TYPE_"
+    edge_id_col_name = "_EDGE_ID_"
+    weight_col_name = "_WEIGHT_"
+    _default_type_name = ""
+
+    def __init__(self, cgs_client, cgs_graph_id):
+        self.__client = cgs_client
+        self.__graph_id = cgs_graph_id
+        self.__vertex_categorical_dtype = None
+        self.__edge_categorical_dtype = None
+
+    def __transform_to_backend_dtype(self, data, column_names, backend, dtypes=[]):
+        """
+        Supports method-by-method selection of backend type (cupy, cudf, etc.)
+        to avoid costly conversion such as row-major to column-major transformation.
+
+        data : cupy.ndarray, np.ndarray
+            The raw ndarray that will be transformed to the backend type.
+        column_names : list[string]
+            The names of the columns, if creating a dataframe.
+        backend : ('cudf', 'cupy') [default = 'cudf']
+            The data backend to convert the provided data to.
+        dtypes : ('int32', 'int64', 'float32', etc.)
+            Optional.  The data type to use when storing data in a dataframe.
+            May be a list, or dictionary corresponding to column names.
+        """
+
+        if backend == "cupy":
+            if isinstance(data, np.ndarray):
+                data = cupy.array(data)
+            return data
+        else:
+            # cudf
+            df = cudf.DataFrame.from_records(data, columns=column_names)
```

Comment:
I think if we are transforming 2d device arrays to cuDF we should do this via `dlpack`  and not via `from_records` . I wonder how this will work if the underlying array is `Row-major`  .

Reply:
resolved; this method no longer handles incoming device arrays